neuralpp/symbolic/general_normalizer.py

```diff
@@ -0,0 +1,121 @@
+    """
+    match expression:
+        case Constant():
+            return expression
+        case FunctionApplication(function=Constant(value=functions.conditional),
+                                 arguments=[condition, then, else_]):
+            if context.is_known_to_imply(condition):
+                return _normalize(then, context)
+            elif context.is_known_to_imply(~condition):
+                return _normalize(else_, context)
+            else:
+                return if_then_else(condition,
+                                    _normalize(then, context & condition),
+                                    _normalize(else_, context & ~condition))
+        case Expression(type=type_) if type_ == bool:
+            if context.is_known_to_imply(expression):
+                return basic_true
+            if context.is_known_to_imply(~expression):
+                return basic_false
+            return if_then_else(expression, True, False)
+        case Variable():
+            return expression
+        case FunctionApplication(function=function, arguments=arguments):
+            return _normalize_function_application(function, arguments, context)
+        case QuantifierExpression(operation=operation, index=index, constraint=constraint, body=body):
+            if context.contains(index):
+                raise ValueError(f"context {context} should not contain index {index}")
+            normalized_body = _normalize(body, context & constraint)
+            match normalized_body:
+                case FunctionApplication(function=Constant(value=functions.conditional),
+                                         arguments=[condition, then, else_]):
+                    if condition.contains(index):
+                        return _normalize(
+                            operation(BasicQuantifierExpression(operation, index, constraint & condition, then),
+                                      BasicQuantifierExpression(operation, index, constraint & ~condition, else_)),
+                            context)
+                    else:
+                        return if_then_else(condition,
+                                            _normalize(BasicQuantifierExpression(operation, index, constraint, then),
+                                                       context & condition),
+                                            _normalize(BasicQuantifierExpression(operation, index, constraint, else_),
+                                                       context & ~condition))
+                case _:
+                    return _eliminate(operation, index, constraint, normalized_body, context)
+
+
+simplifier = ContextSimplifier()
+
+
+def _eliminate(operation: AbelianOperation, index: Variable, constraint: Context, body: Expression,
+               context: Z3SolverExpression) -> Expression:
+    """
+    Eliminates all quantifiers by doing the "summation" (in a broader sense).
```

Comment:
Yes! This might be a better word, though I don't see it used often in the symbolic inference literature (I'm not an expert, so @rodrigodesalvobraz may want to comment on this). I use "sum"/"addition" here because it is the word for operation for [ring](https://en.wikipedia.org/wiki/Ring_(mathematics)) in group theory, from which we took inspiration for some earlier names (e.g., `AbelianOperation`).